crypto/azure/keyvault/algorithms.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keyvault
+
+import (
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/keyvault/azkeys"
+
+	internals "github.com/dapr/kit/crypto"
+)
+
+var (
+	validEncryptionAlgs map[string]struct{}
+	validSignatureAlgs  map[string]struct{}
+)
+
+func init() {
+	listEncryption := azkeys.PossibleJSONWebKeyEncryptionAlgorithmValues()
+	validEncryptionAlgs = make(map[string]struct{}, len(listEncryption))
+	for _, v := range listEncryption {
+		validEncryptionAlgs[string(v)] = struct{}{}
+	}
+
+	listSignature := azkeys.PossibleJSONWebKeySignatureAlgorithmValues()
+	validSignatureAlgs = make(map[string]struct{}, len(listSignature))
+	for _, v := range listSignature {
+		validSignatureAlgs[string(v)] = struct{}{}
+	}
+}
+
+// GetJWKEncryptionAlgorithm returns a JSONWebKeyEncryptionAlgorithm constant is the algorithm is a supported one.
+func GetJWKEncryptionAlgorithm(algorithm string) *azkeys.JSONWebKeyEncryptionAlgorithm {
+	// Special case for AES-CBC, since we treat A[NNN]CBC as having PKCS#7 padding, and A[NNN]CBC-NOPAD as not using padding
+	switch algorithm {
+	case internals.Algorithm_A128CBC, internals.Algorithm_A192CBC, internals.Algorithm_A256CBC:
+		// Append "PAD", e.g. "A128CBCPAD"
+		algorithm += "PAD"
+	case internals.Algorithm_A128CBC_NOPAD, internals.Algorithm_A192CBC_NOPAD, internals.Algorithm_A256CBC_NOPAD:
+		// Remove the "-NOPAD" prefix, e.g. "A128CBC"
```

Comment:
I'll change to suffix.

As for `-NOPAD`, yes it's a constant I made up to smoothen the differences between Azure Key Vault and the RFC 7518 spec (JSON Web Algorithms). 

- In RFC7518, algorithms like `A128CBC` use AES-CBC with PKCS#7 padding (note: PKCS#5 and PKCS#7 padding are essentially equivalent).
  - The RFC doesn't actually define `A128CBC` because the RFC only includes authenticated ciphers. So it actually defines `A128CBC-HS256`, which we will support once https://github.com/dapr/kit/pull/40 is merged. Nevertheless, the specs for `A128CBC-HS256` do require PKCS#7 padding
- Azure Key Vault supports two algorithms (source: https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/keyvault/azkeys#JSONWebKeyEncryptionAlgorithm ):
   - `A128CBC` does **not** use padding
   - `A128CBCPAD` **does** use padding
   - Most people will want to use padding (not using padding means that the plaintext message's length must be a multiple of 16). So for Azure Key Vault we convert `A128CBC` into `A128CBCPAD`, and allow `-NOPAD` algorithms if one really, really doesn't want to use padding